libyul/optimiser/ValueConstraintBasedSimplifier.cpp

```diff
@@ -0,0 +1,384 @@
+		case dev::solidity::Instruction::GT:
+			return visit(args.at(1)) < visit(args.at(0));
+		// TODO SLT, SGT
+		case dev::solidity::Instruction::EQ:
+			return visit(args.at(1)) == visit(args.at(0));
+		case dev::solidity::Instruction::ISZERO:
+			return visit(args.at(0)) == ValueConstraint::constant(0);
+		case dev::solidity::Instruction::AND:
+			return visit(args.at(0)) & visit(args.at(1));
+		case dev::solidity::Instruction::OR:
+			return visit(args.at(0)) | visit(args.at(1));
+		// TODO XOR
+		case dev::solidity::Instruction::NOT:
+			return ~visit(args.at(0));
+		case dev::solidity::Instruction::BYTE:
+			// TODO Could be more specific.
+			return ValueConstraint::bitRange(7, 0);
+		case dev::solidity::Instruction::SHL:
+			return visit(args.at(1)) << visit(args.at(0));
+		case dev::solidity::Instruction::SHR:
+			return visit(args.at(1)) >> visit(args.at(0));
+		// TODO SAR
+		case dev::solidity::Instruction::ADDRESS:
+		case dev::solidity::Instruction::ORIGIN:
+		case dev::solidity::Instruction::CALLER:
+		case dev::solidity::Instruction::COINBASE:
+		case dev::solidity::Instruction::CREATE:
+		case dev::solidity::Instruction::CREATE2:
+			return ValueConstraint::bitRange(159, 0);
+
+		case dev::solidity::Instruction::CALL:
+		case dev::solidity::Instruction::CALLCODE:
+		case dev::solidity::Instruction::DELEGATECALL:
+		case dev::solidity::Instruction::STATICCALL:
+			return ValueConstraint::boolean();
+
+		case dev::solidity::Instruction::KECCAK256:
+		case dev::solidity::Instruction::EXTCODEHASH:
+		case dev::solidity::Instruction::BLOCKHASH:
+		case dev::solidity::Instruction::MLOAD:
+		case dev::solidity::Instruction::SLOAD:
+
+		// These could be restricted in a real-world setting:
+		case dev::solidity::Instruction::BALANCE:
+		case dev::solidity::Instruction::CALLVALUE:
+		case dev::solidity::Instruction::CALLDATASIZE:
+		case dev::solidity::Instruction::GASPRICE:
+		case dev::solidity::Instruction::EXTCODESIZE:
+		case dev::solidity::Instruction::RETURNDATASIZE:
+		case dev::solidity::Instruction::TIMESTAMP:
+		case dev::solidity::Instruction::NUMBER:
+		case dev::solidity::Instruction::DIFFICULTY:
+		case dev::solidity::Instruction::GASLIMIT:
+		case dev::solidity::Instruction::PC:
+		case dev::solidity::Instruction::MSIZE:
+		case dev::solidity::Instruction::GAS:
+		default:
+			return {};
+		}
+	}
+
+private:
+
+	map<YulString, ValueConstraint> const& m_variableConstraints;
+};
+}
+
+ValueConstraint ValueConstraint::constant(u256 const& _value)
+{
+	return ValueConstraint{_value, _value, _value, _value};
+}
+
+ValueConstraint ValueConstraint::boolean()
+{
+	return ValueConstraint{0, 1, 0, 1};
+}
+
+ValueConstraint ValueConstraint::bitRange(size_t _highest, size_t _lowest)
+{
+	yulAssert(_highest >= _lowest, "");
+	return valueFromBits(
+		0,
+		(u256(-1) >> (255 - _highest)) & ~(u256(-1) >> (256 - _lowest))
+	);
+}
+
+ValueConstraint ValueConstraint::valueFromBits(u256 _minBits, u256 _maxBits)
```

Comment:
I think I still didn't understand the difference between values and bits here.
In `valueFromBits`, `maxValue` is `0xff..ff` until the highest bit, even though the mask itself might have zeroes between lowest and highest?
Function `bitsFromValue` is similar, `maxBits` is `0xff..ff` even though `maxValue` might contain zeroes.
Why is it necessary to increase the number when "converting" value -> bits and bits -> value?

Reply:
Answered  as a comment in the main thread.